mrobosub_sim/mrobosub_sim/hal.py

```diff
@@ -203,42 +211,48 @@ def kind(self) -> MessageKind:
 
 
 class SimDepth:
-    def __init__(self) -> None:
-        self.depth_pub = rospy.Publisher("/depth/raw_depth", Float32, queue_size=1)
+    def __init__(self, node: Node) -> None:
+        self.node = node
+        self.depth_pub = self.node.create_publisher(Float32, "/depth/raw_depth", 1)
 
     def handle_sensors(self, data: SensorData):
         self.depth_pub.publish(data.depth)
 
 
 class SimDvl:
-    def __init__(self) -> None:
-        self.dvl_pub = rospy.Publisher("/dvl/raw_dvl", Dvl, queue_size=1)
+    def __init__(self, node: Node) -> None:
+        self.node = node
+        self.dvl_pub = self.node.create_publisher(Dvl, "/dvl/raw_dvl", 1)
 
     def handle_sensors(self, data: SensorData):
         self.dvl_pub.publish(data.dvl)
 
 
 class SimImu:
-    def __init__(self) -> None:
-        self.ins_pub = rospy.Publisher("/imu_INS", Imu_INS, queue_size=1)
-        self.pimu_pub = rospy.Publisher("/imu_PIMU", Imu_PIMU, queue_size=1)
+    def __init__(self, node: Node) -> None:
+        self.node = node
+        self.imu_pub = self.node.create_publisher(Imu, "/imu", 1)
 
     def handle_sensors(self, data: SensorData):
-        self.ins_pub.publish(data.imu_ins)
-        self.pimu_pub.publish(data.imu_pimu)
+        self.imu_pub.publish(data.imu)
 
 
 class SimBotcam:
     def __init__(self, hal: "SimHal") -> None:
```

Comment:
Why do some take a `SimHal` and others take a `Node`?

Reply:
Because some of them call .send on the SimHal and others just need access to create publishers and subscribers